src/parquet/file_reader.cc

```diff
@@ -0,0 +1,764 @@
+#include <thrift/transport/TBufferTransports.h>
+
+#include <iostream>
+#include <sstream>
+#include <limits>
+
+namespace parquet {
+
+////////////////////////////////////////////////////////////////////////////////
+// thrift deserialization
+
+namespace {
+
+template <typename DeserializedType>
+void deserialize_thrift_msg(
+        const uint8_t serialized_msg[],
+        uint32_t* serialized_len,
+        DeserializedType& deserialized_msg
+) {
+    using ThriftBuffer = apache::thrift::transport::TMemoryBuffer;
+    auto tmem_transport = std::make_shared<ThriftBuffer>(const_cast<uint8_t*>(serialized_msg), *serialized_len);
+    apache::thrift::protocol::TCompactProtocolFactoryT<ThriftBuffer> tproto_factory;
+    std::shared_ptr<apache::thrift::protocol::TProtocol> tproto = tproto_factory.getProtocol(tmem_transport);
+    deserialized_msg.read(tproto.get());
+    uint32_t bytes_left = tmem_transport->available_read();
+    *serialized_len = *serialized_len - bytes_left;
+}
+
+} // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+// peekable_stream
+
+/* Assuming there is k bytes remaining in stream, append exactly min(k, n) bytes to the internal buffer.
+ * seastar::input_stream has a read_exactly method of it's own, which does exactly what we want internally,
+ * except instead of returning the k buffered bytes on eof, it discards all of it and returns an empty buffer.
+ * Bummer. */
+seastar::future<> peekable_stream::read_exactly(size_t n) {
+    assert(_buffer.size() - _buffer_end >= n);
+    if (n == 0) {
+        return seastar::make_ready_future<>();
+    }
+    return _source.read_up_to(n).then([this, n] (seastar::temporary_buffer<char> newbuf) {
+        if (newbuf.size() == 0) {
+            return seastar::make_ready_future<>();
+        } else {
+            std::memcpy(_buffer.data() + _buffer_end, newbuf.get(), newbuf.size());
+            _buffer_end += newbuf.size();
+            return read_exactly(n - newbuf.size());
+        }
+    });
+}
+
+/* Ensure that there is at least n bytes of space after _buffer_end.
+ * We want to strike a balance between rewinding the buffer and reallocating it.
+ * If we are too stingy with reallocation, we might do a lot of pointless rewinding.
+ * If we are too stingy with rewinding, we will allocate lots of unused memory too big a buffer.
+ * Case in point: imagine that buffer.size() == 1024.
+ * Then, imagine a peek(1024), advance(1), peek(1024), advance(1)... sequence.
+ * If we never reallocate the buffer, we will have to move 1023 bytes every time we consume a byte.
+ * If we never rewind the buffer, it will keep growing indefinitely, even though we only need 1024 contiguous
+ * bytes.
+ * Our strategy (rewind only when _buffer_start moves past half of buffer.size()) guarantees that
+ * we will actively use at least 1/2 of allocated memory, and that any given byte is rewound at most once.
+ */
+void peekable_stream::ensure_space(size_t n) {
```

Comment:
We talked about unit tests before - that function is one thing that could be unit-tested to ensure that we allocate appropriate amounts of memory for corner cases, just like the one you explained in the comment above.